File: src/main/kotlin/top/bearingwall/asya/service/InstructionService.kt
package top.bearingwall.asya.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import top.bearingwall.asya.audit.Auditable
import top.bearingwall.asya.dto.InstructionCreateRequest
import top.bearingwall.asya.dto.InstructionResponse
import top.bearingwall.asya.dto.InstructionReviewRequest
import top.bearingwall.asya.model.AuditActionType
import top.bearingwall.asya.model.Instruction
import top.bearingwall.asya.model.InstructionStatus
import top.bearingwall.asya.model.InstructionType
import top.bearingwall.asya.model.User
import top.bearingwall.asya.model.UserRole
import top.bearingwall.asya.repository.InstructionRepository
import top.bearingwall.asya.repository.UserRepository
import java.time.LocalDateTime
import java.util.UUID

@Service
class InstructionService(
    private val instructionRepository: InstructionRepository,
    private val userRepository: UserRepository,
    private val timeService: TimeService,
    private val systemConfigService: SystemConfigService
) {

    @Transactional
    @Auditable(type = AuditActionType.INSTRUCTION_CREATE, content = "提交指令")
    fun createInstruction(request: InstructionCreateRequest, submitterUuid: UUID): InstructionResponse {
        val submitter = getUser(submitterUuid)
        require(submitter.role == UserRole.DELEGATE) { "Only DELEGATE can submit instruction" }
        require(!systemConfigService.isInstructionSubmissionPaused()) { "Instruction submission is paused" }

        val conference = submitter.conference ?: throw IllegalStateException("Submitter not associated with any conference")
        val submitRealTime = LocalDateTime.now()
        val submitGameTime = timeService.getCurrentGameTime(conference.uuid!!) ?: submitRealTime

        val saved = instructionRepository.save(
            Instruction(
                conference = conference,
                submitter = submitter,
                title = request.title,
                instructionType = request.instructionType,
                content = request.content,
                submitRealTime = submitRealTime,
                submitGameTime = submitGameTime
            )
        )

        return saved.toResponse()
    }

    @Transactional(readOnly = true)
    fun getMyInstructions(
        submitterUuid: UUID,
        pageable: Pageable,
        status: InstructionStatus? = null
    ): Page<InstructionResponse> {
        getUser(submitterUuid)
        return instructionRepository.findAllBySubmitterUuid(submitterUuid, status, pageable).map { it.toResponse() }
    }

    @Transactional(readOnly = true)
    fun getInstruction(uuid: UUID, requesterUuid: UUID): InstructionResponse {
        val requester = getUser(requesterUuid)
        val instruction = getInstructionEntity(uuid)
        ensureSameConference(requester, instruction)

        val isPrivileged = requester.role in setOf(UserRole.DH, UserRole.DM, UserRole.SYS_ADMIN)
        val isOwner = instruction.submitter.uuid == requester.uuid
        if (!isPrivileged && !isOwner) {
            throw SecurityException("Access denied")
        }

        return instruction.toResponse()
    }

    @Transactional(readOnly = true)
    fun queryInstructionsForManagement(
        requesterUuid: UUID,
        pageable: Pageable,
        status: InstructionStatus?,
        instructionType: InstructionType?,
        userGroupId: Long?,
        submitterUuids: List<UUID>?
    ): Page<InstructionResponse> {
        val requester = getUser(requesterUuid)
        requireManagementRole(requester)
        val conferenceUuid = requester.conference?.uuid ?: throw IllegalStateException("User not associated with any conference")
        val normalizedSubmitterUuids = submitterUuids?.toSet()?.takeIf { it.isNotEmpty() }

        return instructionRepository.findForConferenceManagement(
            conferenceUuid = conferenceUuid,
            status = status,
            instructionType = instructionType,
            userGroupId = userGroupId,
            submitterUuids = normalizedSubmitterUuids,
            pageable = pageable
        ).map { it.toResponse() }
    }

    @Transactional
    @Auditable(type = AuditActionType.INSTRUCTION_REVIEW, content = "批改指令")
    fun reviewInstruction(uuid: UUID, reviewerUuid: UUID, request: InstructionReviewRequest): InstructionResponse {
        val reviewer = getUser(reviewerUuid)
        requireManagementRole(reviewer)
        require(request.status != InstructionStatus.SUBMITTED) { "Review target status cannot be SUBMITTED" }

        val instruction = getInstructionEntity(uuid)
        ensureSameConference(reviewer, instruction)

        val reviewRealTime = LocalDateTime.now()
        val reviewGameTime = timeService.getCurrentGameTime(instruction.conference.uuid!!) ?: reviewRealTime

        instruction.status = request.status
        instruction.reviewComment = request.reviewComment
        instruction.reviewedBy = reviewer
        instruction.reviewedRealTime = reviewRealTime
        instruction.reviewedGameTime = reviewGameTime

        return instructionRepository.save(instruction).toResponse()
    }

    private fun getUser(uuid: UUID): User {
        return userRepository.findById(uuid).orElseThrow {
            IllegalArgumentException("User not found: $uuid")
        }
    }

    private fun getInstructionEntity(uuid: UUID): Instruction {
        return instructionRepository.findById(uuid).orElseThrow {
            IllegalArgumentException("Instruction not found: $uuid")
        }
    }

    private fun requireManagementRole(user: User) {
        if (user.role !in setOf(UserRole.DH, UserRole.DM, UserRole.SYS_ADMIN)) {
            throw SecurityException("Permission denied")
        }
    }

    private fun ensureSameConference(user: User, instruction: Instruction) {
        val userConferenceUuid = user.conference?.uuid
        val instructionConferenceUuid = instruction.conference.uuid
        if (userConferenceUuid == null || userConferenceUuid != instructionConferenceUuid) {
            throw SecurityException("Access denied")
        }
    }

    private fun Instruction.toResponse() = InstructionResponse(
        uuid = this.uuid.toString(),
        conferenceId = this.conference.uuid.toString(),
        submitterId = this.submitter.uuid.toString(),
        submitterName = this.submitter.name,
        title = this.title,
        instructionType = this.instructionType,
        content = this.content,
        status = this.status,
        reviewComment = this.reviewComment,
        submitRealTime = this.submitRealTime,
        submitGameTime = this.submitGameTime,
        reviewedById = this.reviewedBy?.uuid?.toString(),
        reviewedByName = this.reviewedBy?.name,
        reviewedRealTime = this.reviewedRealTime,
        reviewedGameTime = this.reviewedGameTime
    )
}

File: backend-dotnet/src/AsyaMun.Api/Services/InstructionService.cs
using System.Globalization;
using AsyaMun.Api.Data;
using AsyaMun.Api.Dtos;
using AsyaMun.Api.Errors;
using AsyaMun.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace AsyaMun.Api.Services;

public class InstructionService
{
    private const string GameTimeFormat = "yyyy-MM-ddTHH:mm:ss";

    private readonly AppDbContext _db;
    private readonly NotificationService _notificationService;

    public InstructionService(AppDbContext db, NotificationService notificationService)
    {
        _db = db;
        _notificationService = notificationService;
    }

    public async Task<InstructionResponse> CreateInstructionAsync(User submitter, InstructionCreateRequest request)
    {
        if (submitter.Role != UserRole.DELEGATE)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Only DELEGATE can submit instruction");
        }

        var conferenceId = submitter.ConferenceId
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Submitter not associated with any conference");

        var conference = await _db.Conferences.AsNoTracking()
            .FirstOrDefaultAsync(c => c.Uuid == conferenceId)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"Conference not found: {conferenceId}");

        if (conference.InstructionSubmissionPaused)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Instruction submission is paused");
        }

        var submitRealTime = NowUtc();
        var submitGameTime = await GetCurrentGameTimeAsync(conferenceId)
            ?? submitRealTime.ToString(GameTimeFormat, CultureInfo.InvariantCulture);

        var instruction = new Instruction
        {
            Uuid = Guid.NewGuid(),
            ConferenceId = conferenceId,
            SubmitterId = submitter.Uuid,
            Title = request.Title,
            InstructionType = request.InstructionType,
            InstructionContent = request.Content,
            Status = InstructionStatus.SUBMITTED,
            SubmitRealTime = submitRealTime,
            SubmitGameTime = submitGameTime
        };

        await using var transaction = await _db.Database.BeginTransactionAsync();

        _db.Instructions.Add(instruction);
        await _db.SaveChangesAsync();
        await transaction.CommitAsync();

        var persisted = await _db.Instructions.AsNoTracking()
            .Include(i => i.Submitter)
            .Include(i => i.ReviewedByNavigation)
            .FirstAsync(i => i.Uuid == instruction.Uuid);

        return ToResponse(persisted);
    }

    public async Task<Page<InstructionResponse>> GetMyInstructionsAsync(
        User submitter,
        PageInput pageInput,
        InstructionStatus? status,
        string? keyword,
        CancellationToken ct = default)
    {
        IQueryable<Instruction> query = _db.Instructions.AsNoTracking()
            .Include(i => i.Submitter)
            .Include(i => i.ReviewedByNavigation)
            .Where(i => i.SubmitterId == submitter.Uuid);

        if (status.HasValue)
        {
            query = query.Where(i => i.Status == status.Value);
        }

        query = ApplyInstructionKeyword(query, keyword);

        var page = await query.ToPageAsync(ApplyDefaultInstructionSort(pageInput), ct);
        return Page<InstructionResponse>.Of(
            page.Content.Select(ToResponse).ToList(),
            page.Number,
            page.Size,
            page.TotalElements);
    }

    public async Task<InstructionResponse> GetInstructionAsync(Guid uuid, User requester, CancellationToken ct = default)
    {
        var instruction = await _db.Instructions.AsNoTracking()
            .Include(i => i.Submitter)
            .Include(i => i.ReviewedByNavigation)
            .FirstOrDefaultAsync(i => i.Uuid == uuid, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"Instruction not found: {uuid}");

        EnsureSameConference(requester, instruction);

        var isPrivileged = requester.Role is (UserRole.DH or UserRole.DM or UserRole.SYS_ADMIN);
        var isOwner = instruction.SubmitterId == requester.Uuid;
        if (!isPrivileged && !isOwner)
        {
            throw new ForbiddenException("Access denied");
        }

        return ToResponse(instruction);
    }

    public async Task<Page<InstructionResponse>> QueryInstructionsForManagementAsync(
        User requester,
        PageInput pageInput,
        InstructionStatus? status,
        InstructionType? instructionType,
        long? userGroupId,
        IReadOnlyList<Guid>? submitterUuids,
        string? keyword,
        CancellationToken ct = default)
    {
        if (requester.Role is not (UserRole.DH or UserRole.DM or UserRole.SYS_ADMIN))
        {
            throw new ForbiddenException("Permission denied");
        }

        var conferenceId = requester.ConferenceId
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, "User not associated with any conference");

        IQueryable<Instruction> query = _db.Instructions.AsNoTracking()
            .Include(i => i.Submitter)
            .Include(i => i.ReviewedByNavigation)
            .Where(i => i.ConferenceId == conferenceId);

        if (status.HasValue)
        {
            query = query.Where(i => i.Status == status.Value);
        }

        if (instructionType.HasValue)
        {
            query = query.Where(i => i.InstructionType == instructionType.Value);
        }

        if (userGroupId.HasValue)
        {
            var groupId = userGroupId.Value;
            query = query.Where(i => i.Submitter.Groups.Any(g => g.Id == groupId));
        }

        var normalizedSubmitters = submitterUuids?.ToHashSet().Where(id => id != Guid.Empty).ToList();
        if (normalizedSubmitters != null && normalizedSubmitters.Count > 0)
        {
            query = query.Where(i => normalizedSubmitters.Contains(i.SubmitterId));
        }

        query = ApplyInstructionKeyword(query, keyword);

        var page = await query.ToPageAsync(ApplyDefaultInstructionSort(pageInput), ct);
        return Page<InstructionResponse>.Of(
            page.Content.Select(ToResponse).ToList(),
            page.Number,
            page.Size,
            page.TotalElements);
    }

    public async Task<InstructionResponse> ReviewInstructionAsync(Guid uuid, User reviewer, InstructionReviewRequest request)
    {
        if (reviewer.Role is not (UserRole.DH or UserRole.DM or UserRole.SYS_ADMIN))
        {
            throw new ForbiddenException("Permission denied");
        }

        if (request.Status == InstructionStatus.SUBMITTED)
        {
            throw new AsyaBusinessException(BizCode.PARAM_ERROR, "Review target status cannot be SUBMITTED");
        }

        var instruction = await _db.Instructions
            .Include(i => i.Submitter)
            .Include(i => i.ReviewedByNavigation)
            .FirstOrDefaultAsync(i => i.Uuid == uuid)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"Instruction not found: {uuid}");

        EnsureSameConference(reviewer, instruction);

        var reviewRealTime = NowUtc();
        var reviewGameTime = await GetCurrentGameTimeAsync(instruction.ConferenceId)
            ?? reviewRealTime.ToString(GameTimeFormat, CultureInfo.InvariantCulture);

        await using var transaction = await _db.Database.BeginTransactionAsync();

        instruction.Status = request.Status;
        instruction.ReviewComment = request.ReviewComment;
        instruction.ReviewedBy = reviewer.Uuid;
        instruction.ReviewedRealTime = reviewRealTime;
        instruction.ReviewedGameTime = reviewGameTime;

        await _db.SaveChangesAsync();
        await transaction.CommitAsync();

        var persisted = await _db.Instructions.AsNoTracking()
            .Include(i => i.Submitter)
            .Include(i => i.ReviewedByNavigation)
            .FirstAsync(i => i.Uuid == instruction.Uuid);

        if (request.Status == InstructionStatus.FEEDBACKED)
        {
            await _notificationService.NotifyInstructionFeedbackAsync(persisted);
        }

        return ToResponse(persisted);
    }

    public async Task<bool> SetInstructionSubmissionPausedAsync(Guid conferenceId, bool paused, CancellationToken ct = default)
    {
        var conference = await _db.Conferences.FirstOrDefaultAsync(c => c.Uuid == conferenceId, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"Conference not found: {conferenceId}");

        await using var transaction = await _db.Database.BeginTransactionAsync(ct);

        conference.InstructionSubmissionPaused = paused;
        await _db.SaveChangesAsync(ct);
        await transaction.CommitAsync(ct);

        return conference.InstructionSubmissionPaused;
    }

    public async Task<bool> IsInstructionSubmissionPausedAsync(Guid conferenceId, CancellationToken ct = default)
    {
        var conference = await _db.Conferences.AsNoTracking()
            .FirstOrDefaultAsync(c => c.Uuid == conferenceId, ct)
            ?? throw new AsyaBusinessException(BizCode.PARAM_ERROR, $"Conference not found: {conferenceId}");

        return conference.InstructionSubmissionPaused;
    }

    private async Task<string?> GetCurrentGameTimeAsync(Guid conferenceId, CancellationToken ct = default)
    {
        var anchor = await _db.TimeAnchors.AsNoTracking()
            .Where(t => t.ConferenceId == conferenceId && t.IsCurrent)
            .OrderByDescending(t => t.Id)
            .FirstOrDefaultAsync(ct);

        if (anchor == null)
        {
            anchor = await _db.TimeAnchors.AsNoTracking()
                .Where(t => t.ConferenceId == conferenceId)
                .OrderByDescending(t => t.Id)
                .FirstOrDefaultAsync(ct);
        }

        if (anchor?.AnchorGameTime == null)
        {
            return null;
        }

        var ratio = anchor.TimeRatio;
        if (!ratio.HasValue || ratio.Value == 0 || anchor.AnchorRealTime == null ||
            !DateTime.TryParse(anchor.AnchorGameTime, CultureInfo.InvariantCulture, DateTimeStyles.None, out var baseGameTime))
        {
            return anchor.AnchorGameTime;
        }

        var elapsedSeconds = (long)(NowUtc() - anchor.AnchorRealTime.Value).TotalSeconds;
        var gameSeconds = (long)(elapsedSeconds * (double)ratio.Value);

        return FormatGameTime(baseGameTime.AddSeconds(gameSeconds));
    }

    private static IQueryable<Instruction> ApplyInstructionKeyword(IQueryable<Instruction> query, string? keyword)
    {
        var normalized = keyword?.Trim();
        if (string.IsNullOrEmpty(normalized))
        {
            return query;
        }

        var kw = normalized.ToLowerInvariant();
        return query.Where(i =>
            i.Title.ToLower().Contains(kw) ||
            i.InstructionContent.ToLower().Contains(kw));
    }

    private static PageInput ApplyDefaultInstructionSort(PageInput input)
    {
        return input.Sort.Count == 0
            ? new PageInput
            {
                Page = input.Page,
                Size = input.Size,
                Sort = new List<SortSpec>
                {
                    new("SubmitRealTime", true),
                    new("Uuid", true)
                }
            }
            : input;
    }

    private static void EnsureSameConference(User user, Instruction instruction)
    {
        if (user.ConferenceId == null || user.ConferenceId != instruction.ConferenceId)
        {
            throw new ForbiddenException("Access denied");
        }
    }

    private static InstructionResponse ToResponse(Instruction instruction)
    {
        return new InstructionResponse(
            instruction.Uuid.ToString("D"),
            instruction.ConferenceId.ToString("D"),
            instruction.SubmitterId.ToString("D"),
            instruction.Submitter.Name,
            instruction.Title,
            instruction.InstructionType,
            instruction.InstructionContent,
            instruction.Status,
            instruction.ReviewComment,
            instruction.SubmitRealTime,
            instruction.SubmitGameTime,
            instruction.ReviewedBy?.ToString("D"),
            instruction.ReviewedByNavigation?.Name,
            instruction.ReviewedRealTime,
            instruction.ReviewedGameTime);
    }

    private static DateTime NowUtc()
    {
        return DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified);
    }

    private static string FormatGameTime(DateTime time)
    {
        var baseValue = time.ToString("yyyy-MM-ddTHH:mm", CultureInfo.InvariantCulture);
        if (time.Second == 0 && time.Millisecond == 0)
        {
            return baseValue;
        }

        if (time.Millisecond == 0)
        {
            return baseValue + ":" + time.Second.ToString("00", CultureInfo.InvariantCulture);
        }

        var fraction = time.Millisecond.ToString("fff", CultureInfo.InvariantCulture).TrimEnd('0');
        return baseValue + ":" + time.Second.ToString("00", CultureInfo.InvariantCulture) + "." + fraction;
    }
}